### PR TITLE
Nktchinh

### PR DESCRIPTION
**Pull request by ntkchinh**

### CODE REVIEW

File: assets/js/linear_algebra/dot_product_collide.js [hunk 1]
```diff
@@ -15,17 +15,17 @@ let origin = [150, 140],
...
+    last_row_coord = start_coord_y + 0.775 * h_unit;
 
-let bot_right_x = start_coord_x + dws_array.reduce(
-    function(total, d) {return total + d * w_unit;}, 0),
-    bot_right_y = start_coord_y + dhs_array.reduce(
-    function(total, d) {return total + d * h_unit;}, 0);
+let v_cell = {text: 'v =', x: last_col_coord,
+              y: start_coord_y - 0.7 * h_unit, key: 'v'},
+    uT_cell = {text: 'u\u1d40 =', x: start_coord_x - 0.6 * w_unit,
+               y: last_row_coord + 0.09 * h_unit, key: 'u' };
```
Comment (thtrieu) [inline assets/js/linear_algebra/dot_product_collide.js]: can we have a variable for these constants 0.7, 0.6, 0.09

File: assets/js/linear_algebra/dot_product_collide.js [hunk 2]
```diff
@@ -176,7 +173,8 @@ function plot(scatter, axis, tt){
        text_color: 0, text_opacity: uTvv_opacity,
        key: 'uTv_texts'},
       {text: uTv.toFixed(3), text_color: 0, 
-       x: bot_right_x, y: bot_right_y,
+       x: last_col_coord,
+       y: last_row_coord + 0.09 * h_unit,
```
Comment (thtrieu) [inline assets/js/linear_algebra/dot_product_collide.js]: name for 0.09?

File: assets/js/linear_algebra/dot_product_collide.js [hunk 3]
```diff
@@ -332,8 +328,9 @@ function compute(u, v){
   ];
   
   for (i = 0; i < animation_end.length; i++) {
-      animation_end[i].x = bot_right_x;
-      animation_end[i].y = bot_right_y;
+      animation_end[i].x = last_col_coord;
+      animation_end[i].y = last_row_coord + 0.09 * h_unit;
```
Comment (thtrieu) [inline assets/js/linear_algebra/dot_product_collide.js]: name for 0.09

File: assets/js/linear_algebra/dot_product_collide.js [hunk 4]
```diff
@@ -346,7 +343,8 @@ function compute(u, v){
        text_color: 0, text_opacity: 1.0,
        key: 'uTv_texts'},
       {text: uTv.toFixed(3), text_color: 0, 
-       x: bot_right_x, y: bot_right_y,
+       x: last_col_coord,
+       y: last_row_coord + 0.09 * h_unit,
```
Comment (thtrieu) [inline assets/js/linear_algebra/dot_product_collide.js]: name for 0.09

File: assets/js/linear_algebra/dot_product_collide2d.js [hunk 5]
```diff
@@ -14,18 +14,21 @@ let origin = [150, 140],
...
+    last_row_coord = start_coord_y + 0.6 * h_unit;
+
+let v_cell = {text: 'v =', x: last_col_coord,
+              y: start_coord_y - 0.4 * h_unit, key: 'v'},
+    u_cell = {text: 'u =', x: last_col_coord,
+              y: start_coord_y - 0.4 * h_unit, key: 'u'},
+    vT_cell = {text: 'v\u1d40 =', x: start_coord_x - 0.6 * w_unit,
+               y: last_row_coord + 0.09, key: 'v' },
+    uT_cell = {text: 'u\u1d40 =', x: start_coord_x - 0.6 * w_unit,
+               y: last_row_coord + 0.09, key: 'u' };
```
Comment (thtrieu) [inline assets/js/linear_algebra/dot_product_collide2d.js]: name for 0.4, 0.6 and 0.09?

File: assets/js/linear_algebra/dot_product_collide2d.js [hunk 6]
```diff
@@ -134,34 +137,38 @@ function plot(scatter, axis, tt){
...
+          );
+
+  let texts_to_show = [],
+      lines_to_show = [];
+
+  let zv_text = {text: '', x: last_col_coord,
+                 y: start_coord_y + 0.3 * h_unit, text_opacity: 0, key: 'zv'
+                },
+      zuT_text = {text: '', x: start_coord_x + 0.6 * w_unit,
+                  y: last_row_coord + 0.09 * h_unit, text_opacity: 0, key: 'zu'
```
Comment (thtrieu) [inline assets/js/linear_algebra/dot_product_collide2d.js]: name for 0.3 and 0.09

File: assets/js/linear_algebra/dot_product_collide2d.js [hunk 7]
```diff
@@ -175,8 +182,10 @@ function plot(scatter, axis, tt){
        y: v.y * uTv/2,
        text_color: 0, text_opacity: uTvv_opacity,
        tt: tt, key: 'uTv_texts'},
-      {text: uTv.toFixed(3), text_color: 0, 
-       x: bot_right_x, y: bot_right_y,
+      {text: uTv.toFixed(3), text_color: 0,
+       x: last_col_coord,
+       y: last_row_coord + 0.09 * h_unit,
+       // x: bot_right_x, y: bot_right_y,
```
Comment (thtrieu) [inline assets/js/linear_algebra/dot_product_collide2d.js]: rm

File: assets/js/linear_algebra/dot_product_collide2d.js [hunk 8]
```diff
@@ -175,8 +182,10 @@ function plot(scatter, axis, tt){
        y: v.y * uTv/2,
        text_color: 0, text_opacity: uTvv_opacity,
        tt: tt, key: 'uTv_texts'},
-      {text: uTv.toFixed(3), text_color: 0, 
-       x: bot_right_x, y: bot_right_y,
+      {text: uTv.toFixed(3), text_color: 0,
+       x: last_col_coord,
+       y: last_row_coord + 0.09 * h_unit,
```
Comment (thtrieu) [inline assets/js/linear_algebra/dot_product_collide2d.js]: 0.09 name

File: assets/js/linear_algebra/dot_product_collide2d.js [hunk 9]
```diff
@@ -293,48 +302,42 @@ function compute(u, v){
...
           {text: u.coord.y.toFixed(2), text_opacity: 0, tt: 450, key: 'yuu'},
   ];
   for (i = 0; i < animation_end.length; i++) {
-      animation_end[i].x = bot_right_x;
-      animation_end[i].y = bot_right_y;
+      animation_end[i].x = last_col_coord;
+      animation_end[i].y = last_row_coord + 0.09 * h_unit;
   };
   
   lib.plot_texts(animation_end, tt=0, name='animation');
```
Comment (thtrieu) [inline assets/js/linear_algebra/dot_product_collide2d.js]: fix this argument name, no argument name passing in js

File: assets/js/linear_algebra/dot_product_formula.js [hunk 10]
```diff
@@ -128,39 +129,39 @@ function plot(scatter, axis, tt){
...
+                   y: end_coord_y + 0.12 * h_unit, z: 0},
+                  {x: start_coord_x + 1.6 * w_unit,
+                   y: end_coord_y + 0.12 * h_unit, z: 0,
                    color: 0}],
       uTv_texts = [{text: 'u\u1d40v =', x: start_coord_x - 0.6 * w_unit,
-                    y: start_coord_y + 0.9 * h_unit, text_color: 0, key: 'uTv_text'},
+                    y: end_coord_y + 0.37 * h_unit, text_color: 0, key: 'uTv_text'},
                    {text: uTv.toFixed(3), x: start_coord_x + 0.4 * w_unit,
-                    y: start_coord_y + 0.9, text_color: 0, key: 'uTv_numb'},];
+                    y: end_coord_y + 0.37 * h_unit, text_color: 0, key: 'uTv_numb'},];
```
Comment (thtrieu) [inline assets/js/linear_algebra/dot_product_formula.js]: name for some of the constants will help with reading the code.

File: assets/js/linear_algebra/dot_product_formula.js [hunk 11]
```diff
@@ -177,6 +178,7 @@ function plot(scatter, axis, tt){
   lines_to_show.push(...lines_u);
   lines_to_show.push(...lines_v);
   lines_to_show.push(big_line);
+  console.log(lines_to_show);
```
Comment (thtrieu) [inline assets/js/linear_algebra/dot_product_formula.js]: rm

File: assets/js/linear_algebra/dot_product_formula2d.js [hunk 12]
```diff
@@ -17,6 +17,7 @@ let w_unit = 1.0, h_unit = 1.0;
 
 let start_coord_x=(400 - origin[0])/scale,
     start_coord_y=(125 - origin[1])/scale;
+    end_coord_y = start_coord_y + 0.33 * h_unit;
 
 let u_cell = {text: 'u =', x: start_coord_x,
               y: start_coord_y - 0.45 * h_unit, key: 'u'},
```
Comment (thtrieu) [inline assets/js/linear_algebra/dot_product_formula2d.js]: name for constant.

File: assets/js/linear_algebra/lib.js [hunk 13]
```diff
@@ -776,72 +776,81 @@ function text_table_to_list(texts,
...
+      side_margin = 0.03 * size/14,
+      between_rows = 0.12 * size/14,
+      w_col = w_unit * size/14,
+      h_row = h_unit * size/14,
+      bracket_head = 0.1 * size/14,
+
+      matrix_w = side_margin * 2 + w_col * numb_of_vector,
+
+      matrix_h = bot_margin + h_row * size_of_space +
+                  between_rows * (size_of_space - 1),
```
Comment (thtrieu) [inline assets/js/linear_algebra/lib.js]: great names!

File: assets/js/linear_algebra/dot_product_collide.js (hunk 1 above)
Comment (ntkchinh) [reply assets/js/linear_algebra/dot_product_collide.js]: done

File: assets/js/linear_algebra/dot_product_collide.js (hunk 2 above)
Comment (ntkchinh) [reply assets/js/linear_algebra/dot_product_collide.js]: done

File: assets/js/linear_algebra/dot_product_collide.js (hunk 3 above)
Comment (ntkchinh) [reply assets/js/linear_algebra/dot_product_collide.js]: done

File: assets/js/linear_algebra/dot_product_collide.js (hunk 4 above)
Comment (ntkchinh) [reply assets/js/linear_algebra/dot_product_collide.js]: done

File: assets/js/linear_algebra/dot_product_collide2d.js (hunk 5 above)
Comment (ntkchinh) [reply assets/js/linear_algebra/dot_product_collide2d.js]: done

File: assets/js/linear_algebra/dot_product_collide2d.js (hunk 9 above)
Comment (ntkchinh) [reply assets/js/linear_algebra/dot_product_collide2d.js]: done

File: assets/js/linear_algebra/dot_product_formula.js (hunk 10 above)
Comment (ntkchinh) [reply assets/js/linear_algebra/dot_product_formula.js]: done

File: assets/js/linear_algebra/dot_product_formula.js (hunk 11 above)
Comment (ntkchinh) [reply assets/js/linear_algebra/dot_product_formula.js]: done

File: assets/js/linear_algebra/dot_product_formula2d.js (hunk 12 above)
Comment (ntkchinh) [reply assets/js/linear_algebra/dot_product_formula2d.js]: done

File: assets/js/linear_algebra/lib.js (hunk 13 above)
Comment (ntkchinh) [reply assets/js/linear_algebra/lib.js]: thanks

**Comment by thtrieu** [comment]
Ok, all clear. thanks!

**Pull request merged**